File: python/ghleak/reporter.py
"""Results reporting — JSON, CSV, terminal table, markdown."""

from __future__ import annotations

import csv
import json
import logging
import textwrap
from collections import Counter
from datetime import datetime, timezone
from typing import Any, TextIO

from .scanner import TruffleHogResult

log = logging.getLogger(__name__)


class Report:
    """Aggregate report for a pipeline run."""

    def __init__(self) -> None:
        self.start_time = datetime.now(timezone.utc)
        self.end_time: datetime | None = None
        self.hours_scanned: int = 0
        self.commits_seen: int = 0
        self.suspicious_commits: list[dict[str, Any]] = []
        self.verified_findings: list[TruffleHogResult] = []
        self.classifier_stats: dict[str, int] = {}
        self.errors: list[str] = []

    def add_suspicious(self, repo: str, sha: str, message: str, classification: str) -> None:
        self.suspicious_commits.append({
            "repo": repo,
            "sha": sha,
            "message": message,
            "classification": classification,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        })

    def add_finding(self, finding: TruffleHogResult) -> None:
        self.verified_findings.append(finding)

    def finish(self) -> None:
        self.end_time = datetime.now(timezone.utc)

    @property
    def duration(self) -> str:
        if not self.end_time:
            return "running"
        delta = self.end_time - self.start_time
        return f"{delta.total_seconds():.1f}s"

    @property
    def verified_count(self) -> int:
        return sum(1 for f in self.verified_findings if f.verified)

    def summary_text(self, width: int = 72) -> str:
        lines = [
            "═" * width,
            "  GhLeak — Secret Scan Report",
            "═" * width,
            f"  Scanned hours : {self.hours_scanned}",
            f"  Commits seen  : {self.commits_seen:,}",
            f"  Suspicious    : {len(self.suspicious_commits)}",
            f"  Verified      : {self.verified_count}",
            f"  Total findings: {len(self.verified_findings)}",
            f"  Duration      : {self.duration}",
        ]

        if self.classifier_stats:
            lines.append("")
            lines.append("  Classifier stats:")
            for k, v in self.classifier_stats.items():
                lines.append(f"    {k}: {v}")

        if self.verified_findings:
            lines.append("")
            lines.append(f"  Top detectors (verified):")
            detector_counts = Counter(
                f.detector for f in self.verified_findings if f.verified
            )
            for detector, count in detector_counts.most_common(10):
                lines.append(f"    {detector}: {count}")

            lines.append("")
            lines.append("  Sample verified findings:")
            for f in self.verified_findings[:5]:
                if f.verified:
                    lines.append(
                        f"    [{f.detector}] {f.repository} @ {f.commit[:12]}"
                        f"  ({f.file})"
                    )

        if self.errors:
            lines.append("")
            lines.append(f"  Errors ({len(self.errors)}):")
            for err in self.errors[:5]:
                lines.append(f"    ! {err}")

        lines.append("═" * width)
        return "\n".join(lines)

    def to_dict(self) -> dict[str, Any]:
        return {
            "start_time": self.start_time.isoformat(),
            "end_time": self.end_time.isoformat() if self.end_time else None,
            "duration": self.duration,
            "hours_scanned": self.hours_scanned,
            "commits_seen": self.commits_seen,
            "suspicious_count": len(self.suspicious_commits),
            "verified_count": self.verified_count,
            "total_findings": len(self.verified_findings),
            "classifier_stats": self.classifier_stats,
            "suspicious_commits": self.suspicious_commits,
            "verified_findings": [
                {
                    "detector": f.detector,
                    "verified": f.verified,
                    "commit": f.commit,
                    "file": f.file,
                    "repository": f.repository,
                    "email": f.email,
                    "redacted": f.redacted,
                    "raw": f.raw_v2[:120] + "..." if len(f.raw_v2) > 120 else f.raw_v2,
                }
                for f in self.verified_findings
            ],
            "errors": self.errors,
        }

    def to_json(self, file: TextIO | None = None) -> str:
        data = self.to_dict()
        out = json.dumps(data, indent=2, default=str)
        if file:
            file.write(out)
        return out

    def to_csv(self, file: TextIO) -> None:
        writer = csv.writer(file)
        writer.writerow(["detector", "verified", "commit", "file", "repository", "email"])
        for f in self.verified_findings:
            writer.writerow([f.detector, f.verified, f.commit, f.file, f.repository, f.email])

    def to_markdown(self) -> str:
        """Render as GitHub-flavored markdown (like the blog post style)."""
        lines = [
            "# GhLeak Scan Report",
            "",
            f"**Scan period:** {self.start_time.isoformat()} — {self.end_time.isoformat() if self.end_time else 'running'}",
            f"**Duration:** {self.duration}",
            f"**Hours scanned:** {self.hours_scanned}",
            f"**Commits seen:** {self.commits_seen:,}",
            f"**Suspicious commits flagged:** {len(self.suspicious_commits)}",
            f"**Verified secrets found:** {self.verified_count}",
            f"**Total findings:** {len(self.verified_findings)}",
            "",
        ]

        if self.verified_findings:
            lines.append("## Verified Findings by Detector")
            lines.append("")
            lines.append("| Detector | Count |")
            lines.append("|----------|-------|")
            detector_counts = Counter(
                f.detector for f in self.verified_findings if f.verified
            )
            for detector, count in detector_counts.most_common():
                lines.append(f"| {detector} | {count} |")

            lines.append("")
            lines.append("## Sample Verified Secrets")
            lines.append("")
            lines.append("| Detector | Repository | Commit | File |")
            lines.append("|----------|------------|--------|------|")
            for f in self.verified_findings[:20]:
                if f.verified:
                    lines.append(
                        f"| {f.detector} | {f.repository or '—'} | "
                        f"`{f.commit[:12]}` | {f.file or '—'} |"
                    )

        if self.suspicious_commits:
            lines.append("")
            lines.append("## Suspicious Commits (by message classification)")
            lines.append("")
            lines.append("| Repository | Commit | Message |")
            lines.append("|------------|--------|---------|")
            for c in self.suspicious_commits[:30]:
                msg = textwrap.shorten(c["message"], width=60, placeholder="...")
                lines.append(
                    f"| {c['repo']} | `{c['sha'][:12]}` | {msg} |"
                )

        return "\n".join(lines)

File: go/internal/reporter.go
package internal

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
	"time"
)

// NewReport creates an empty thread-safe report.
func NewReport() *Report {
	return &Report{
		SuspiciousCommits: []SuspiciousCommit{},
		Findings:          []Finding{},
		ClassifierStats:   map[string]int{},
		Errors:            []string{},
		StartTime:         time.Now().UTC().Format(time.RFC3339),
	}
}

// IncHoursScanned safely increments the hours counter.
func (r *Report) IncHoursScanned() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.HoursScanned++
}

// IncCommitsSeen safely increments the commits counter.
func (r *Report) IncCommitsSeen() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.CommitsSeen++
}

// AddSuspicious appends a flagged commit to the report securely.
func (r *Report) AddSuspicious(repo, sha, message, result string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.SuspiciousCommits = append(r.SuspiciousCommits, SuspiciousCommit{
		Repo: repo, SHA: sha, Message: message, Result: result,
	})
}

// AddFinding appends a TruffleHog finding securely.
func (r *Report) AddFinding(f Finding) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.Findings = append(r.Findings, f)
	if f.Verified {
		r.VerifiedCount++
	}
}

// AddError records a pipeline error.
func (r *Report) AddError(err string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.Errors = append(r.Errors, err)
}

// Finish records the end time.
func (r *Report) Finish() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.EndTime = time.Now().UTC().Format(time.RFC3339)
}

// Duration returns the elapsed time between start and end as a formatted string.
func (r *Report) Duration() string {
	r.mu.Lock()
	defer r.mu.Unlock()
	start, err := time.Parse(time.RFC3339, r.StartTime)
	if err != nil {
		return "unknown"
	}
	end, err := time.Parse(time.RFC3339, r.EndTime)
	if err != nil {
		return "unknown"
	}
	d := end.Sub(start)
	return fmt.Sprintf("%.0fs", d.Seconds())
}

// SummaryText returns a human-readable summary matching Python's format.
func (r *Report) SummaryText() string {
	r.mu.Lock()
	defer r.mu.Unlock()

	var b strings.Builder
	b.WriteString("════════════════════════════════════════════════════════════════════════\n")
	b.WriteString("  GhLeak — Secret Scan Report\n")
	b.WriteString("════════════════════════════════════════════════════════════════════════\n")
	b.WriteString(fmt.Sprintf("  Scanned hours : %d\n", r.HoursScanned))
	b.WriteString(fmt.Sprintf("  Commits seen  : %d\n", r.CommitsSeen))
	b.WriteString(fmt.Sprintf("  Suspicious    : %d\n", len(r.SuspiciousCommits)))
	b.WriteString(fmt.Sprintf("  Verified      : %d\n", r.VerifiedCount))
	b.WriteString(fmt.Sprintf("  Total findings: %d\n", len(r.Findings)))

	if len(r.SuspiciousCommits) > 0 {
		b.WriteString("\n  Suspicious Commits:\n")
		b.WriteString(fmt.Sprintf("  %-20s | %-12s | %s\n", "Repository", "SHA", "Message"))
		b.WriteString(fmt.Sprintf("  %s\n", strings.Repeat("-", 72)))
		for _, sc := range r.SuspiciousCommits {
			sha := sc.SHA
			if len(sha) > 12 {
				sha = sha[:12]
			}
			msg := sc.Message
			if len(msg) > 50 {
				msg = msg[:50] + "..."
			}
			b.WriteString(fmt.Sprintf("  %-20s | %-12s | %s\n", sc.Repo, sha, msg))
		}
	}

	if len(r.Findings) > 0 {
		b.WriteString(fmt.Sprintf("\n  Top detectors (verified):\n"))
		detectorCounts := map[string]int{}
		for _, f := range r.Findings {
			if f.Verified {
				detectorCounts[f.DetectorName]++
			}
		}
		type dc struct {
			name  string
			count int
		}
		sorted := make([]dc, 0, len(detectorCounts))
		for name, count := range detectorCounts {
			sorted = append(sorted, dc{name, count})
		}
		sort.Slice(sorted, func(i, j int) bool {
			return sorted[i].count > sorted[j].count
		})
		for i, d := range sorted {
			if i >= 10 {
				break
			}
			b.WriteString(fmt.Sprintf("    %s: %d\n", d.name, d.count))
		}

		b.WriteString("\n  Sample verified findings:\n")
		sampleCount := 0
		for _, f := range r.Findings {
			if sampleCount >= 5 {
				break
			}
			if f.Verified {
				commit := f.Commit
				if len(commit) > 12 {
					commit = commit[:12]
				}
				b.WriteString(fmt.Sprintf("    [%s] %s @ %s  (%s)\n",
					f.DetectorName, f.Repository, commit, f.File))
				sampleCount++
			}
		}
	}

	if len(r.ClassifierStats) > 0 {
		b.WriteString("\n  Classifier stats:\n")
		for k, v := range r.ClassifierStats {
			b.WriteString(fmt.Sprintf("    %s: %d\n", k, v))
		}
	}

	if len(r.Errors) > 0 {
		b.WriteString(fmt.Sprintf("\n  Errors (%d):\n", len(r.Errors)))
		for i, err := range r.Errors {
			if i >= 5 {
				break
			}
			b.WriteString(fmt.Sprintf("    ! %s\n", err))
		}
	}

	b.WriteString("════════════════════════════════════════════════════════════════════════\n")
	return b.String()
}

// ToJSON serializes the report as indented JSON with transformed findings.
func (r *Report) ToJSON() string {
	r.mu.Lock()
	defer r.mu.Unlock()

	transformedFindings := make([]map[string]any, 0, len(r.Findings))
	for _, f := range r.Findings {
		raw := f.RawV2
		if len(raw) > 120 {
			raw = raw[:120] + "..."
		}
		transformedFindings = append(transformedFindings, map[string]any{
			"detector":   f.DetectorName,
			"verified":   f.Verified,
			"commit":     f.Commit,
			"file":       f.File,
			"repository": f.Repository,
			"email":      f.Email,
			"redacted":   f.Redacted,
			"raw":        raw,
		})
	}

	out := map[string]any{
		"start_time":         r.StartTime,
		"end_time":           r.EndTime,
		"hours_scanned":      r.HoursScanned,
		"commits_seen":       r.CommitsSeen,
		"suspicious_count":   len(r.SuspiciousCommits),
		"verified_count":     r.VerifiedCount,
		"total_findings":     len(r.Findings),
		"classifier_stats":   r.ClassifierStats,
		"suspicious_commits": r.SuspiciousCommits,
		"verified_findings":  transformedFindings,
		"errors":             r.Errors,
	}

	data, _ := json.MarshalIndent(out, "", "  ")
	return string(data)
}

// ToCSV writes findings to a CSV writer matching Python's columns.
func (r *Report) ToCSV(w io.Writer) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	cw := csv.NewWriter(w)
	cw.Write([]string{"detector", "verified", "commit", "file", "repository", "email"})
	for _, f := range r.Findings {
		cw.Write([]string{
			f.DetectorName,
			fmt.Sprintf("%v", f.Verified),
			f.Commit,
			f.File,
			f.Repository,
			f.Email,
		})
	}
	cw.Flush()
	return cw.Error()
}

// ToMarkdown formats the report as Markdown matching Python's format.
func (r *Report) ToMarkdown() string {
	r.mu.Lock()
	defer r.mu.Unlock()

	var b strings.Builder
	b.WriteString("# GhLeak Scan Report\n\n")

	b.WriteString(fmt.Sprintf("**Hours scanned:** %d\n", r.HoursScanned))
	b.WriteString(fmt.Sprintf("**Commits seen:** %d\n", r.CommitsSeen))
	b.WriteString(fmt.Sprintf("**Suspicious:** %d\n", len(r.SuspiciousCommits)))
	b.WriteString(fmt.Sprintf("**Verified secrets:** %d\n\n", r.VerifiedCount))

	if len(r.Findings) > 0 {
		b.WriteString("## Verified Findings by Detector\n\n")
		b.WriteString("| Detector | Count |\n")
		b.WriteString("|----------|-------|\n")
		detectorCounts := map[string]int{}
		for _, f := range r.Findings {
			if f.Verified {
				detectorCounts[f.DetectorName]++
			}
		}
		type dc struct {
			name  string
			count int
		}
		sorted := make([]dc, 0, len(detectorCounts))
		for name, count := range detectorCounts {
			sorted = append(sorted, dc{name, count})
		}
		sort.Slice(sorted, func(i, j int) bool {
			return sorted[i].count > sorted[j].count
		})
		for _, d := range sorted {
			b.WriteString(fmt.Sprintf("| %s | %d |\n", d.name, d.count))
		}

		b.WriteString("\n## Sample Verified Secrets\n\n")
		b.WriteString("| Detector | Repository | Commit | File |\n")
		b.WriteString("|----------|------------|--------|------|\n")
		sampleCount := 0
		for _, f := range r.Findings {
			if sampleCount >= 20 {
				break
			}
			if f.Verified {
				repo := f.Repository
				if repo == "" {
					repo = "—"
				}
				file := f.File
				if file == "" {
					file = "—"
				}
				commit := f.Commit
				if len(commit) > 12 {
					commit = commit[:12]
				}
				b.WriteString(fmt.Sprintf("| %s | %s | `%s` | %s |\n",
					f.DetectorName, repo, commit, file))
				sampleCount++
			}
		}
	}

	if len(r.SuspiciousCommits) > 0 {
		b.WriteString("\n## Suspicious Commits (by message classification)\n\n")
		b.WriteString("| Repository | Commit | Message |\n")
		b.WriteString("|------------|--------|---------|\n")
		for i, sc := range r.SuspiciousCommits {
			if i >= 30 {
				break
			}
			sha := sc.SHA
			if len(sha) > 12 {
				sha = sha[:12]
			}
			msg := sc.Message
			if len(msg) > 60 {
				msg = msg[:60] + "..."
			}
			b.WriteString(fmt.Sprintf("| %s | `%s` | %s |\n", sc.Repo, sha, msg))
		}
	}

	return b.String()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// OutputReport handles formatting and writing the report.
func OutputReport(r *Report, format string, outFile string) {
	var output string

	switch format {
	case "json":
		output = r.ToJSON()
	case "csv":
		if outFile != "" {
			f, err := os.Create(outFile)
			if err == nil {
				r.ToCSV(f)
				f.Close()
				fmt.Printf("Report written to %s\n", outFile)
				return
			}
		}
		r.ToCSV(os.Stdout)
		return
	case "md":
		output = r.ToMarkdown()
	default:
		output = r.SummaryText()
	}

	if outFile != "" {
		os.WriteFile(outFile, []byte(output), 0644)
		fmt.Printf("Report written to %s\n", outFile)
	} else {
		fmt.Println(output)
	}
}
